internal/tracker/downstream.go

```diff
@@ -0,0 +1,51 @@
+package tracker
+
+import (
+	"sync"
+)
+
+// DownstreamConns tracks connections per downstream based on a
+// unique string identifier.
+// DownstreamConns safe for concurrent use.
+type DownstreamConns struct {
+	// mu protects the resources of DownstreamConns
+	mu sync.Mutex
+
+	// connCounts is a map of downstream to a count of connections
+	connCounts map[string]uint32
+}
+
+// NewDownstreamConns initializes and returns a DownstreamConns with
+func NewDownstreamConns() *DownstreamConns {
```

Comment:
If the constructor takes no arguments, how can you pass the available backends?

Reply:
Ahhh, great catch. I had left that as a mental TODO, but hadn't implemented the call to NewDownstreamConns() yet, so I had yet to revisit this bit. I'll fix this.

Reply:
Actually, scratch that. I misread this and was responding to your comment as if it was concerning upstreams.

downstreamConns has no notion of available backends, it only tracks client connections. the conns map starts empty but is populated as new clients request connections.